libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/targets/native/tasks/KotlinNativePerformanceTask.kt

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.gradle.targets.native.tasks
+
+import org.gradle.api.DefaultTask
+import org.gradle.api.Task
+import org.gradle.api.tasks.*
+import org.jetbrains.kotlin.gradle.plugin.KotlinCompilation
+import org.jetbrains.kotlin.gradle.plugin.mpp.NativeBinary
+import org.jetbrains.kotlin.gradle.plugin.performance.PerformanceExtension
+import org.jetbrains.kotlin.gradle.plugin.performance.TaskTimerListener
+import org.jetbrains.kotlin.gradle.plugin.performance.TrackableMetric
+import java.io.File
+
+/**
+ * The task generates performance report for Kotlin/Native binary.
+ */
+open class NativePerformanceReport : DefaultTask() {
+    @Internal
+    lateinit var binary: NativeBinary
+
+    @Internal
+    lateinit var timeListener: TaskTimerListener
+
+    @Internal
+    val reportDirectory = File(project.buildDir, "perfReports")
+
+    @OutputFile
+    val outputFile = File(reportDirectory, "${name}.txt")
+
+    @Internal
+    lateinit var settings: PerformanceExtension
+
+    private fun getCompilationResults(tasksPathes: Iterable<String>, success: Boolean): String {
+        val status = success && tasksPathes.all { timeListener.tasksTimes.containsKey(it) }
+        val time = tasksPathes.map { timeListener.getTime(it) }.sum()
+        return "${if (status) "PASSED" else "FAILED"}\nCOMPILE_TIME $time"
+    }
+
+    // Get compile task and associated with it other compile tasks.
+    private fun getAllExecutedTasks(compilation: KotlinCompilation<*>): List<Task> {
+        val tasks = mutableListOf(compilation.compileKotlinTask as Task)
+        compilation.associateWith.forEach {
+            tasks += getAllExecutedTasks(it)
+        }
+        return tasks
+    }
+
+    @TaskAction
+    fun generate() {
+        val compileTasks = if (settings.includeAssociatedTasks)
+            getAllExecutedTasks(binary.linkTask.compilation)
+        else
+            listOf(binary.linkTask.compilation.compileKotlinTask)
+        val allExecutedTasks = listOf(binary.linkTask) + compileTasks
+        val upToDateTasks = allExecutedTasks.filter { it.state.upToDate }.map { it.name }.joinToString("\n", "- ")
+        if (upToDateTasks.isEmpty()) {
+            if (reportDirectory.exists()) {
+                project.delete(reportDirectory.absolutePath)
```

Comment:
Not sure if we need to delete the whole report directory here.

Reply:
I think, we need, because then user can skip this warning and use wrong report. If user already run new build he's ready that this directory inside build will be rewritten.